Azure/azure-dev
# ado with oidc

fix: https://github.com/Azure/azure-dev/issues/2777

Updating azd pipeline config to support Federated Credential for Azure DevOps

The azdo provider's order of configuration was updated to depend on the `auth-type` like:

## federated

- First azd creates a Service Principal in EntraId w/o any credential (not federated or client-secret)
- Then the azdo provider creates a service-connection using the SP details (clientId and TenantId) using:
  - `endpointScheme := "WorkloadIdentityFederation"` 
- Once the service connection is created, azd takes the `Issuer` and `Subject` from it to create the federated credential for the SP.

## client-credentials

- First azd creates a Service Principal in EntraId w/o any credential (not federated or client-secret)
- Then azd creates a secret-client credential for the SP
- Then the azdo provider creates a service-connection using the SP details (clientId, secret and TenantId) using:
  - `endpointScheme = "ServicePrincipal"`


## Manual tests

For each test, use a new fresh azd project which has not local git project. For example, run:
```
dotnet new aspire-starter
azd init
azd pipeline config <flags>
```

Then, make sure there is a `.azdo/pipelines` folder on the root of the project with a `azure-dev.yml`:
![image](https://github.com/Azure/azure-dev/assets/24213737/c81e1db4-b03d-403c-bda8-1b85e795d29b)

Here is a sample for the pipeline definition for the sample app:
```yaml
trigger:
  - main
  - master
pool:
  vmImage: ubuntu-latest
steps:
  - task: setup-azd@0 
    displayName: Install azd
  - pwsh: |
      azd config set auth.useAzCliAuth "true"
    displayName: Configure AZD to Use AZ CLI Authentication.
  - pwsh: |
      dotnet workload restore
    displayName: restore
  - task: AzureCLI@2
    displayName: Provision Infrastructure
    inputs:
      azureSubscription: azconnection
      scriptType: bash
      scriptLocation: inlineScript
      inlineScript: |
        azd provision --no-prompt
    env:
      AZURE_SUBSCRIPTION_ID: $(AZURE_SUBSCRIPTION_ID)
      AZURE_ENV_NAME: $(AZURE_ENV_NAME)
      AZURE_LOCATION: $(AZURE_LOCATION)
  - task: AzureCLI@2
    displayName: Deploy Application
    inputs:
      azureSubscription: azconnection
      scriptType: bash
      scriptLocation: inlineScript
      inlineScript: |
        azd deploy --no-prompt
    env:
      AZURE_SUBSCRIPTION_ID: $(AZURE_SUBSCRIPTION_ID)
      AZURE_ENV_NAME: $(AZURE_ENV_NAME)
      AZURE_LOCATION: $(AZURE_LOCATION)
```

### base tests

- [x] ado with fedCred  *use no flags should use federated by default  // azd pipeline config --provider azdo
- [x] ado with fedCred  *use flags for federated // azd pipeline config --provider azdo --auth-type federated
- [x] azd with clientSecret *use flags for client-secret // azd pipeline config --provider azdo --auth-type client-credentials

### Migration test

- [x] set up clientSecret, then update it to fedCred.  Expectation:  The SP would end up having 2 credentials (one with secret and another using fedCredential).  The service connection in Azdo should be using the fedCredential.
- [x] set up fedCred, then update it to clientSecret. Expectation:  The SP would end up having 2 credentials (one with secret and another using fedCredential).  The service connection in Azdo should be using the clientSecret.

### Regression test

Repeat all test above using github as provider

- [x] 2
- [x] 3
- [x] 4
- [x] 5
